File: R/normalize_to_one.R
# Identify the index of the element with the largest absolute difference from its original value
# Calculate the difference between 1 and the sum of the normalized vector
# Add the difference to the element with the largest absolute difference

normalize_to_one <- function(x) {
  if (!is.numeric(x)) {
    stop("Input must be a numeric vector")
  }
  normalized_x <- x/sum(x)
  
  diff <- 1-sum(normalized_x)
  
  max_diff_idx <- which.max(abs(normalized_x - x))
  normalized_x[max_diff_idx] <- normalized_x[max_diff_idx] + diff
  
  return(normalized_x)
}

File: Python/normalize_to_one.py
# Identify the index of the element with the largest absolute difference from its original value
# Add the difference to the element with the largest absolute difference
def normalize_to_one(x):
    if not all(isinstance(i, (int, float)) for i in x):
        raise ValueError("Input must be a list or tuple of numerics")

    normalized_x = [i / sum(x) for i in x]
    diff = 1 - sum(normalized_x)

    max_diff_idx = max(range(len(x)), key=lambda i: abs(normalized_x[i] - x[i]))
    normalized_x[max_diff_idx] += diff

    return normalized_x